### PR TITLE
Add refresh index API for partition skipping index

### PR DESCRIPTION
**Pull request by dai-chen**
### Description

Please see more details in changes in doc: https://github.com/dai-chen/sql-1/blob/add-partion-index-building/flint/docs/index.md

#### Flint Metadata Spec

Update metadata with more info (index kind, column name and type). In this way, `FlintSparkSkippingIndex` can be deserialized out of `_meta` JSON.

#### Flint Spark API

Add `refreshIndex` API with `FULL` (batch) and `INCREMENTAL` (streaming) mode.

#### Partition Index Implementation

Followed the first approach below for now. Need to evaluate the performance later.

- Approach 1: treat partition columns as normal and use `FIRST_VALUE` function to aggregate
- Approach 2: treat partition columns as special and add partition columns to `GROUP BY`

```
scala> spark.sql("select first_value(year, true) as year, first_value(month, true) as month, first_value(day, true), first_value(hour, true), input_file_name() from alb_logs group by input_file_name()").explain
== Physical Plan ==
*(2) HashAggregate(keys=[_nondeterministic#146], functions=[first_value(year#33, true), first_value(month#34, true), first_value(day#35, true), first_value(hour#36, true)])
+- Exchange hashpartitioning(_nondeterministic#146, 200), ENSURE_REQUIREMENTS, [id=#112]
   +- *(1) HashAggregate(keys=[_nondeterministic#146], functions=[partial_first_value(year#33, true), partial_first_value(month#34, true), partial_first_value(day#35, true), partial_first_value(hour#36, true)])
      +- *(1) Project [year#33, month#34, day#35, hour#36, input_file_name() AS _nondeterministic#146]
         +- FileScan csv default.alb_logs[year#33,month#34,day#35,hour#36] Batched: false, DataFilters: [], Format: CSV, Location: CatalogFileIndex[s3a://maximus.dev.daichen.us-west-2/alb-log-daily], PartitionFilters: [], PushedFilters: [], ReadSchema: struct<>

scala> spark.sql("select year, month, day, hour, input_file_name() from alb_logs group by year, month, day, hour, input_file_name()").explain
== Physical Plan ==
*(2) HashAggregate(keys=[year#33, month#34, day#35, hour#36, _nondeterministic#169], functions=[])
+- Exchange hashpartitioning(year#33, month#34, day#35, hour#36, _nondeterministic#169, 200), ENSURE_REQUIREMENTS, [id=#134]
   +- *(1) HashAggregate(keys=[year#33, month#34, day#35, hour#36, _nondeterministic#169], functions=[])
      +- *(1) Project [year#33, month#34, day#35, hour#36, input_file_name() AS _nondeterministic#169]
         +- FileScan csv default.alb_logs[year#33,month#34,day#35,hour#36] Batched: false, DataFilters: [], Format: CSV, Location: CatalogFileIndex[s3a://maximus.dev.daichen.us-west-2/alb-log-daily], PartitionFilters: [], PushedFilters: [], ReadSchema: struct<>
```

Sample index data:

```
curl "localhost:9200/flint_test_skipping_index/_search?pretty"
{
  "took" : 121,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 18,
      "relation" : "eq"
    },
    "max_score" : 1.0,
    "hits" : [
      {
        "_index" : "flint_test_skipping_index",
        "_id" : "LWt3VIgBQCGf4ClzQDd8",
        "_score" : 1.0,
        "_source" : {
          "file_path" : "file:///.../spark-warehouse/test/year=2023/month=4/part-00000-c2f88187-da00-40aa-981a-471df2e15ca1.c000.csv",
          "year" : 2023,
          "month" : 4
        }
      },
      {
        "_index" : "flint_test_skipping_index",
        "_id" : "L2t6VIgBQCGf4ClzVzcU",
        "_score" : 1.0,
        "_source" : {
          "file_path" : "file:///.../spark-warehouse/test/year=2023/month=5/part-00000-10f5c696-8fd8-4b28-8d41-62ac9c39f905.c000.csv",
          "year" : 2023,
          "month" : 5
        }
      }
      ...
```
 
### Issues Resolved

https://github.com/opensearch-project/sql/issues/1602
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: flint/integ-test/src/test/scala/org/opensearch/flint/spark/FlintSparkSkippingIndexSuite.scala [hunk 1]
```diff
@@ -93,6 +124,56 @@ class FlintSparkSkippingIndexSuite extends FlintSuite with OpenSearchSuite {
...
+
+    val job = spark.streams.get(jobId.get)
+    failAfter(streamingTimeout) {
+      job.processAllAvailable()
+    }
+
+    val indexData =
+      spark.read
+        .format(FLINT_DATASOURCE)
+        .schema("year INT, month INT, file_path STRING")
```
Comment (penghuo) [inline flint/integ-test/src/test/scala/org/opensearch/flint/spark/FlintSparkSkippingIndexSuite.scala:169]: another way to provide schema? good to know.

File: flint/flint-spark-integration/src/main/scala/org/opensearch/flint/spark/skipping/partition/PartitionSkippingStrategy.scala [hunk 2]
```diff
@@ -7,14 +7,24 @@ package org.opensearch.flint.spark.skipping.partition
...
-class PartitionSkippingStrategy(override val indexedColumn: (String, String))
+class PartitionSkippingStrategy(
+    override val kind: String = "partition",
+    override val columnName: String,
+    override val columnType: String)
     extends FlintSparkSkippingStrategy {
 
   override def outputSchema(): Map[String, String] = {
-    Map(indexedColumn._1 -> convertToFlintType(indexedColumn._2))
+    Map(columnName -> convertToFlintType(columnType))
```
Comment (penghuo) [inline flint/flint-spark-integration/src/main/scala/org/opensearch/flint/spark/skipping/partition/PartitionSkippingStrategy.scala:23]: does the flintType align need to align with AggregateFunction output type?

**Review by penghuo — APPROVED** [review]

File: flint/flint-spark-integration/src/main/scala/org/opensearch/flint/spark/skipping/FlintSparkSkippingStrategy.scala [hunk 3]
```diff
@@ -5,19 +5,33 @@
...
+   * Skipping strategy kind.
+   */
+  val kind: String
+
   /**
    * Indexed column name and its Spark SQL type.
    */
-  val indexedColumn: (String, String)
+  val columnName: String
+  val columnType: String
```
[the remark this answers is not in the document]
Comment (dai-chen) [reply flint/flint-spark-integration/src/main/scala/org/opensearch/flint/spark/skipping/FlintSparkSkippingStrategy.scala:24]: sure, I may remove this `columnType` field from `_meta` and this class after we refactored the Spark SQL type => Flint type mapping

File: flint/flint-spark-integration/src/main/scala/org/opensearch/flint/spark/skipping/partition/PartitionSkippingStrategy.scala (hunk 2 above)
Comment (dai-chen) [reply flint/flint-spark-integration/src/main/scala/org/opensearch/flint/spark/skipping/partition/PartitionSkippingStrategy.scala:23]: yes, it needs to align. As comment above, I may need to use AggregateFunction output type if `columnType` is unnecessary. I've noted this down. Thanks!

**Pull request merged**